src/frontend/test_runner/tests/testdata/range_scan.yaml

```diff
@@ -33,17 +33,13 @@
   sql: |
     SELECT * FROM orders_count_by_user WHERE user_id = 'a'
   batch_plan: |
-    BatchExchange { order: [], dist: Single }
-      BatchFilter { predicate: false:Boolean }
-        BatchScan { table: orders_count_by_user, columns: [user_id, date, orders_count] }
+    BatchValues { rows: [] }
```

Comment:
Hmmm... Why? `user_id = 'a'` is always NULL?

Reply:
FYI. PG reports error in this case:

```
dev=# select * from orders where o_custkey = 'a';
ERROR:  invalid input syntax for type integer: "a"
LINE 1: select * from orders where o_custkey = 'a';
```

Reply:
fixed

Reply:
Actually not completely fixed.. Currently it will only return cast error when building scan range.

This case which fails in PG will not fail in risingwave (it's runtime error), unless we introduce constant folding:
```sql
create table t(x int);

select * from t where x='a'; -- ERROR:  invalid input syntax for type integer: "a"

select x='a' from t; -- ERROR:  invalid input syntax for type integer: "a"
```

Reply:
What's your idea? cc @xiangjinwu

Reply:
I am aware of this difference, but do not have an issue tracking it. 😢 `'a'` is a literal of `unknown` type. After its type has been determined by the type inference rules, we wrap it with a cast expr and eval in backend, while pg does the parsing in "frontend" directly.

https://github.com/singularity-data/risingwave/blob/7f0ff39c3dec093e0603471462d5939c321795d9/src/frontend/test_runner/tests/testdata/expr.yaml#L127-L133

To match the pg behavior, we could leverage the new `eval_row_const`. However, I think the better way is to impl `FromStr` for all `ScalarImpl` variants rather than relying on this general expr eval. This requires some refactor of cast-from-string implementation.

Reply:
Given that a complete solution to match PG would happen in binder:
https://github.com/singularity-data/risingwave/blob/f690d5ef18cc91c4fe0e4e230223212117ddae92/src/frontend/src/expr/function_call.rs#L181-L194

The handling of eval_**cast** error is unnecessary, as **implicit cast (excluding parsing `unknown`) never produce runtime error** (correct me if I am wrong).

But there may be other errors during eval_**const**, for example:
`id_int32 in (2147483647 + 1)`

So I guess the introduction of `Result` still makes sense. It should not become `false` silently.

Reply:
> But there may be other errors during eval_const, for example:
id_int32 in (2147483647 + 1) (maybe add to test case?)

Yes, I already added `SELECT * FROM orders_count_by_user WHERE user_id = 1/0`

Reply:
I also considered about improve cast with `eval_const` in this PR, but I think it's still not enough, e.g., this cannot be handled:

```sql
select * from t where x=1/0; -- ERROR:  division by zero
```
🤔

Reply:
To summarize: 
1. const expr error should be handled in frontend?
2. cast_implicit can itself handle parsing. BTW, if we have 1, this seems uncessary (but maybe still good to have) 🤔